Codegen arange.start_out

wconstab opened the pull request:
- does not add full support for factory functions (with TensorOption
  style args), but does add codegen support for handling expanding
  TensorOption args into their components.  Needs more work in order
  to support ops like `arange`
- supports arange.start_out: changes to how shape inference
  functions are named, to accomodate having out variants vs functional variants

RUN_TORCHBENCH: ALL
TORCHBENCH_BRANCH: wconstab/ltc

alanwaketan reviewed (commented):
Not seeing the op being added to code-gen?

alanwaketan reviewed (approved):
LGTM!

wconstab commented:
oops this breaks.  I guess I hadn't saved the .yaml when I thought I was testing codegen of arange.  We need to land the fix for this first @alanwaketan 

AssertionError: need to initialize local.use_const_ref_for_mutable_tensors with local.parametrize

wconstab committed:
Add codegen support for arange.start_out

- does not add full support for factory functions (with TensorOption
  style args), but does add codegen support for handling expanding
  TensorOption args into their components.  Needs more work in order
  to support ops like `arange`
- supports arange.start_out: needs changes to how shape inference
  functions are named

WIP:
- shape inference for arange.start_out is wrong
- Arange test fails, which is odd since I expected only arange_out to be
  impacted

wconstab committed:
Fix codegen handling of out= variant of arange, return out tensor

wconstab committed:
Use is_out_fn() rather than stringmatch

wconstab committed:
make arange.start_out use correct shape inference

wconstab committed:
Fix bug where size_d was shadowed

alanwaketan reviewed (approved):
LGTM.

wconstab commented on lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeInference.cpp:
```diff
@@ -76,6 +77,58 @@ std::vector<int64_t> expand_param_if_needed(
+  double size_d;
+  // shape inference code copied from RangeFactories.cpp arange_out function
+  // Note: AT_DISPATCH_ALL_TYPES_AND is just a macro that defines the correct c++ scalar_t type depending on out tensor
+  AT_DISPATCH_ALL_TYPES_AND(c10::kBFloat16, out.scalar_type(), "compute_shape_arange_out", [&]() {
```
this macro literally does nothing other than define scalar_t to be a c++ type that is appropriate given out.scalar_type().  

out.scalar_type() is an enum, with a value such as 'float_type', and given this value, scalar_type would be set to `float`.

wconstab commented on lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeInference.cpp:
```diff
@@ -76,6 +77,58 @@ std::vector<int64_t> expand_param_if_needed(
+  // Note: AT_DISPATCH_ALL_TYPES_AND is just a macro that defines the correct c++ scalar_t type depending on out tensor
+  AT_DISPATCH_ALL_TYPES_AND(c10::kBFloat16, out.scalar_type(), "compute_shape_arange_out", [&]() {
+    // Note: acc_type further defines an accumulataion type depending on the scalar_t and whether its on cuda vs cpu.
+    using accscalar_t = at::acc_type<scalar_t, false>;
```
acc_type encodes the few special cases where a larger type than scalar_t is used for accumulation.  This differs between cpu and cuda, so this function returns the right type depending on both scalar_t and device.
